registry/storage/cache/cachedblobdescriptorstore.go

```diff
@@ -53,47 +28,37 @@ func NewCachedBlobStatter(cache distribution.BlobDescriptorService, backend dist
 	}
 }
 
-// NewCachedBlobStatterWithMetrics creates a new statter which prefers a cache and
-// falls back to a backend. Hits and misses will send to the tracker.
-func NewCachedBlobStatterWithMetrics(cache distribution.BlobDescriptorService, backend distribution.BlobDescriptorService, tracker MetricsTracker) distribution.BlobStatter {
-	return &cachedBlobStatter{
-		cache:   cache,
-		backend: backend,
-		tracker: tracker,
-	}
-}
-
 func (cbds *cachedBlobStatter) Stat(ctx context.Context, dgst digest.Digest) (distribution.Descriptor, error) {
 	cacheCount.WithValues("Request").Inc(1)
-	desc, err := cbds.cache.Stat(ctx, dgst)
-	if err != nil {
-		if err != distribution.ErrBlobUnknown {
-			logErrorf(ctx, cbds.tracker, "error retrieving descriptor from cache: %v", err)
-		}
 
-		goto fallback
-	}
-	cacheCount.WithValues("Hit").Inc(1)
-	if cbds.tracker != nil {
-		cbds.tracker.Hit()
-	}
-	return desc, nil
-fallback:
-	cacheCount.WithValues("Miss").Inc(1)
-	if cbds.tracker != nil {
-		cbds.tracker.Miss()
+	// try getting from cache
+	desc, cacheErr := cbds.cache.Stat(ctx, dgst)
+	if cacheErr == nil {
+		cacheCount.WithValues("Hit").Inc(1)
+		return desc, nil
 	}
-	desc, err = cbds.backend.Stat(ctx, dgst)
+
+	// couldn't get from cache; get from backend
+	desc, err := cbds.backend.Stat(ctx, dgst)
 	if err != nil {
 		return desc, err
 	}
 
-	if err := cbds.cache.SetDescriptor(ctx, dgst, desc); err != nil {
-		logErrorf(ctx, cbds.tracker, "error adding descriptor %v to cache: %v", desc.Digest, err)
+	if cacheErr == distribution.ErrBlobUnknown {
+		// cache doesn't have info. update it with info got from backend
+		cacheCount.WithValues("Miss").Inc(1)
+		if err := cbds.cache.SetDescriptor(ctx, dgst, desc); err != nil {
+			dcontext.GetLoggerWithField(ctx, "blob", dgst).WithError(err).Error("error from cache setting desc")
+		}
+		// we don't need to return cache error upstream if any. continue returning value from backend
+		return desc, nil
 	}
 
-	return desc, err
+	// unknown error from cache. just log and error. do not store cache as it may be trigger many set calls
+	dcontext.GetLoggerWithField(ctx, "blob", dgst).WithError(cacheErr).Error("error from cache stat(ing) blob")
+	cacheCount.WithValues("Error").Inc(1)
 
+	return desc, nil
```

Comment:
Minor nit: This return can be once. So above can be written:
```go
	if cacheErr == distribution.ErrBlobUnknown {
		// cache doesn't have info. update it with info got from backend
		cacheCount.WithValues("Miss").Inc(1)
		if err := cbds.cache.SetDescriptor(ctx, dgst, desc); err != nil {
			dcontext.GetLoggerWithField(ctx, "blob", dgst).WithError(err).Error("error from cache setting desc")
		}
	} else {
	    // unknown error from cache. just log and error. do not store cache as it may be trigger many set calls
	    dcontext.GetLoggerWithField(ctx, "blob", dgst).WithError(cacheErr).Error("error from cache stat(ing) blob")
	   cacheCount.WithValues("Error").Inc(1)
       }
       return desc, nil
````